### PR TITLE
feat: add support for allay platform

### PR DESCRIPTION
**Pull request by smartcmd**
# Pull Request

## Description

This PR add support for [allay](https://github.com/AllayMC/Allay) platform.

### Changelog

- [x] Basic support
    - [x] Block & item & biome mappings
- [ ] ~~Command delegation~~
- [x] Reloading support
- [ ] ~~Mapping generator~~

#### Mandatory checks

- [ ] The base branch of this PR is an unreleased version branch (that has a `ver/` prefix)
  or is a branch that is intended to be merged into a version branch.
  
- [x] There are no already existing PRs that provide the same changes.
  
- [x] The PR is within the scope of Terra (i.e. is something a configurable terrain generator should be doing).
- [x] Changes follow the code style for this project.
  
- [x] I have read the [`CONTRIBUTING.md`](https://github.com/PolyhedralDev/Terra/blob/master/CONTRIBUTING.md)
  document in the root of the git repository.

#### Types of changes

- [ ] Bug Fix 
- [ ] Build system 
- [ ] Documentation 
- [x] New Feature 
- [ ] Performance 
- [ ] Refactoring 
- [ ] Repository 
- [ ] Revert 
- [ ] Style 
- [ ] Tests 
- [ ] Translation 

#### Compatibility

- [ ] Introduces a breaking change
  
- [ ] Introduces new functionality in a backwards compatible way.
- [ ] Introduces bug fixes

#### Documentation

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

#### Testing

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
  

#### Licensing

- [x] I am the original author of this code, and I am willing to
  release it under [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html).
- [ ] I am not the original author of this code, but it is in public domain or
  released under [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html) or a compatible license.

### CODE REVIEW

File: platforms/allay/build.gradle.kts
Comment (duplexsystem) [inline platforms/allay/build.gradle.kts:6]: These should be added to the DependencyConfig.kt file instead of here

File: platforms/allay/src/main/java/org/allaymc/terra/allay/Mapping.java
Comment (duplexsystem) [inline platforms/allay/src/main/java/org/allaymc/terra/allay/Mapping.java:24]: If we have to use lombok I'd prefer if we moved to `private static final Logger LOGGER = LoggerFactory.getLogger(Class.class);` for consistency with the rest of the code base

File: platforms/allay/src/main/java/org/allaymc/terra/allay/AllayPlatform.java
Comment (duplexsystem) [inline platforms/allay/src/main/java/org/allaymc/terra/allay/AllayPlatform.java:1]: this should be com.dfsek.terra.allay

File: platforms/allay/build.gradle.kts
Comment (duplexsystem) [inline platforms/allay/build.gradle.kts:16]: All of these versions should go in Versions.kt. Gson should go to Libraries internal as I plan to use it for other things and it's a good library to have set there. everything else should go in a a new allay section. Also is lombok necessary as part of the platform or is it just a convenience if it's not required I'd prefer if we moved away from it here as to not introduce any giant dependencies. Also we should pin a version of allay so a new version does not have to be downloaded every time you open the an ide to work on Terra.
Also follow the format (package, dep, version) like gson.

File: platforms/allay/src/main/java/org/allaymc/terra/allay/generator/AllayGeneratorWrapper.java
Comment (duplexsystem) [inline platforms/allay/src/main/java/org/allaymc/terra/allay/generator/AllayGeneratorWrapper.java:28]: Same as before

File: platforms/allay/src/main/java/org/allaymc/terra/allay/TerraAllayPlugin.java
Comment (duplexsystem) [inline platforms/allay/src/main/java/org/allaymc/terra/allay/TerraAllayPlugin.java:13]: Same as before

File: platforms/allay/src/main/java/org/allaymc/terra/allay/generator/AllayGeneratorWrapper.java
Comment (duplexsystem) [inline platforms/allay/src/main/java/org/allaymc/terra/allay/generator/AllayGeneratorWrapper.java:31]: For consistency with other platforms I'd prefer if Terra was disabled if the pack is not set

**Review by duplexsystem — CHANGES_REQUESTED** [review]
Looks good except for a few things:

- Are we excepting the TODOs to be done before merge?
- Is it possible to not use coding styles like 'var' for consistency with the rest of the code base?
- If Allay is trying to emulate minecraft (which it is) i'd recommend implementing the additional platform biome options using fabric as a reference see (https://github.com/PolyhedralDev/Terra/tree/master/platforms/mixin-common/src/main/java/com/dfsek/terra/mod/config) for consistency with other platforms.

Over all good work!

File: platforms/allay/README.md
Comment (duplexsystem) [inline platforms/allay/README.md:8]: The mapping json files should not have version info in the file names. they should just be 'mapping/biomes.json' instead of 'mapping/biomes_JE_1_21_to_BE_1_21_30.json' for example.

**Comment by smartcmd** [comment]
> Looks good except for a few things:
> 
> * Are we excepting the TODOs to be done before merge?
> * Is it possible to not use coding styles like 'var' for consistency with the rest of the code base?
> * If Allay is trying to emulate minecraft (which it is) i'd recommend implementing the additional platform biome options using fabric as a reference see (https://github.com/PolyhedralDev/Terra/tree/master/platforms/mixin-common/src/main/java/com/dfsek/terra/mod/config) for consistency with other platforms.
> 
> Over all good work!

* All 'var' have been replaced with the explicit type
* I'm not going to implement the remaining two features (command, additional biome options) as they are a bit difficult to implement now c:. However, config pack reloading should be easy to be done and I will implement it.

**Comment by duplexsystem** [comment]
> > Looks good except for a few things:
> > 
> > * Are we excepting the TODOs to be done before merge?
> > * Is it possible to not use coding styles like 'var' for consistency with the rest of the code base?
> > * If Allay is trying to emulate minecraft (which it is) i'd recommend implementing the additional platform biome options using fabric as a reference see (https://github.com/PolyhedralDev/Terra/tree/master/platforms/mixin-common/src/main/java/com/dfsek/terra/mod/config) for consistency with other platforms.
> > 
> > Over all good work!
> 
>     * All 'var' have been replaced with the explicit type
> 
>     * I'm not going to implement the remaining two features (command, additional biome options) as they are a bit difficult to implement now c:. However, config pack reloading should be easy to be done and I will implement it.

Sounds good.

**Comment by duplexsystem** [comment]
Is it possible to pin allay to a specific build instead of using the latest build for compiling?

**Comment by smartcmd** [comment]
> Is it possible to pin allay to a specific build instead of using the latest build for compiling?

image

Yes, I already did this c:

**Comment by duplexsystem** [comment]
Ah sorry my git window had not refreshed since commit [8a6ad95](https://github.com/PolyhedralDev/Terra/pull/474/commits/8a6ad9594799cb4c40764b39ee793fba27efe193)

**Comment by smartcmd** [comment]
Is there anything else that needs to be done?

**Comment by duplexsystem** [comment]
No it will get merged in the next release cycle

**Comment by solonovamax** [comment]
should @smartcmd be added as a CODEOWNER for the allay platform?

**Comment by duplexsystem** [comment]
Yes I think, we should also add @OakLoaf as a code owner for bukkit

**Pull request merged**
